### ISSUE TITLE
Plan with -detailed-exitcode option returns error 

### ISSUE THREAD

**Issue by aminzar-access** [hashicorp/setup-terraform#152]
Hi, 
Running plan with -detailed-exitcode reports error. Thanks.

![image](https://user-images.githubusercontent.com/77307027/156144298-a7bcee4b-7baf-443d-85d3-79a56341e199.png)

```HCL
      - name: Terraform Plan
        id: plan
        run: terraform -chdir=${{ env.working_directory}} plan -refresh=${{ github.event_name != 'pull_request' }} -out ${{ env.tf_plan_file }} -detailed-exitcode -no-color
        continue-on-error: true
```

**Comment by Michaelvsk** [comment]
That's not related to `setup-terraform` action. [Github Actions handle all nonzero exit codes as errors](https://docs.github.com/en/actions/creating-actions/setting-exit-codes-for-actions) and therefore let the script end immediately and prints an error. 

If you do not need to process the detailed exit code, then just omit the parameter. If you want to do different things based on the exit code, then you are in the same situation as I am currently. So far I haven't found a workaround to retrieve the detailed exit code but do not let the script stop immedialty.

**Comment by culpinnis** [comment]
Hi together,
We used a snippet like this before to determine if the plan contained any changes. 
```bash
          set +e
          terraform plan ${{ inputs.TF_VARS }} -detailed-exitcode -out=deploy_${{ github.run_number }}.tfplan ${{ steps.SetTerraformDestroyFlag.outputs.destroyFlag }}
          exitcode=$?
          if [[ $exitcode -eq 2 ]]; then
            echo '::set-output name=planHasChanges::true'
            exit 0
          else
            exit $exitcode
          fi
```
This works fine when using an ubuntu runner without the TF setup action. 
When using the provided Hashicorp action with the following code, I get the same issue. 
```yaml
      - uses: hashicorp/setup-terraform@v1
        name: Setup Terraform on private runner
```
Of course, I can use `|| true`, but then I do not know if the plan has any changes (the error is reported to the log, but the exit code is 0 instead of 2). 
Any ideas on how to deal with that? Is this a general GitHub action issue or related to this specific action?

**Comment by aereal** [comment]
This is an issue that comes from @actions/core function.

hashicorp/setup-terraform provides the terraform(1) wrapper and it uses `@actions/core.setFailed` function.
refs. https://github.com/hashicorp/setup-terraform/blob/8b4c280fc8c755f3640ce104b5ba443608256909/index.js#L9

`@actions/core.setFailed` always sets exit code as `1` and discards `terraform plan`'s exit code such as `2`.
refs. https://github.com/actions/toolkit/blob/main/packages/core/src/core.ts#L213-L217

We can obtain correct exit code from terraform(1) to specify `terraform_wrapper: false` and use bare terraform CLI.

**Comment by aminzar-access** [comment]
> This is an issue that comes from @actions/core function.
> 
> hashicorp/setup-terraform provides the terraform(1) wrapper and it uses `@actions/core.setFailed` function. refs.
> 
> https://github.com/hashicorp/setup-terraform/blob/8b4c280fc8c755f3640ce104b5ba443608256909/index.js#L9
> 
> `@actions/core.setFailed` always sets exit code as `1` and discards `terraform plan`'s exit code such as `2`. refs. https://github.com/actions/toolkit/blob/main/packages/core/src/core.ts#L213-L217
> 
> We can obtain correct exit code from terraform(1) to specify `terraform_wrapper: false` and use bare terraform CLI.

Hi @aereal, how do we get `stdout`, `stderr` and the `exitcode` with `terraform_wrapper: false`? I can't find any samples. Thanks

**Comment by aereal** [comment]
@aminzar-access 

You may be able to set those values using [workflow commands](https://docs.github.com/en/actions/using-workflows/workflow-commands-for-github-actions#setting-an-output-parameter).

```shell
steps:
  - id: terraform
    run: |
      set +e # allow the workflow to continue following steps even if the `terraform` execution failed
      terraform ... 1>stdout 2>stderr
      tf_exitcode=$?
      echo "::set-output name=exitcode::$tf_exitcode"
      echo "::set-output name=stdout::$(cat ./stdout)"
      echo "::set-output name=stderr::$(cat ./stderr)"
```

(I'm not sure using `set-output` with multi-line text works correctly)

**Comment by aminzar-access** [comment]
Thanks for your help @aereal, really appreciate it.

**Comment by aminzar-access** [comment]
@aereal unfortunately seems like `exitcode` is not being set and `stdout` doesn't have a value when you have multiple terraform commands in workflow.

**Comment by jpogran** [comment]
We recently merged https://github.com/hashicorp/setup-terraform/pull/125, which handles both 0 and 2 as exitcodes. If you could test by using that git commit I would sincerely appreciate it. An official tag will be created when [2.1.0](https://github.com/hashicorp/setup-terraform/milestone/3) is released

**Comment by aminzar-access** [comment]
@jpogran looks like it has solved the issue, no more errors and I can still get the proper exitcode, thanks.

**Comment by brettcurtis** [comment]
Hoping for 2.1.0 soon too, this is really confusing to our users.

**Comment by ianhundere** [comment]
+1

**Comment by austinvalle** [comment]
Closing this issue as it seems the fix was released in [`v2.0.1`](https://github.com/hashicorp/setup-terraform/releases/tag/v2.0.1) 👍🏻
